### ISSUE TITLE
bazel failed to build with Xcode 9.0 

### ISSUE THREAD

**Issue by zhexuany** [bazelbuild/bazel#3787]
The log is the following:
~~~
ERROR: /private/var/tmp/_bazel_zhexuany/d90cfb1d83b5fac0d2a4fea11702e03d/external/com_github_google_protobuf/BUILD:73:1: undeclared inclusion(s) in rule '@com_github_google_protobuf//:protobuf_lite':
this rule is missing dependency declarations for the following files included by 'external/com_github_google_protobuf/src/google/protobuf/arena.cc':
  '/Library/Developer/CommandLineTools/usr/lib/clang/9.0.0/include/stddef.h'
  '/Library/Developer/CommandLineTools/usr/lib/clang/9.0.0/include/__stddef_max_align_t.h'
  '/Library/Developer/CommandLineTools/usr/lib/clang/9.0.0/include/stdint.h'
  '/Library/Developer/CommandLineTools/usr/lib/clang/9.0.0/include/stdarg.h'
  '/Library/Developer/CommandLineTools/usr/lib/clang/9.0.0/include/limits.h'
  '/Library/Developer/CommandLineTools/usr/lib/clang/9.0.0/include/stdbool.h'.
INFO: Elapsed time: 0.835s, Critical Path: 0.46s
~~~

**Comment by damienmg** [comment]
Hi,

Can you give us the step you use to build Bazel? This looks like the C++ compiler has changed.

**Comment by jgavris** [comment]
I've also seen this...happens every time we upgrade our toolchain. `bazel clean --expunge` will get you back in order, but it would be better if we could gracefully upgrade toolchains.

**Comment by davidcairns** [comment]
I just ran into this as well; changing toolchains via `xcode-select --switch` requires doing a manual `bazel shutdown`, then a `bazel clean --expunge` (h/t to @jgavris for that suggestion!).

**Comment by girving** [comment]
Happening to me about after upgrading to the Xcode 10 beta.  Any results of the September 2017 investigation?  `bazel shutdown` and `bazel clean --expunge` didn't work.  The error after those when building tensorflow is

```
INFO: Analysed target //tensorflow/tools/pip_package:build_pip_package (332 packages loaded).
INFO: Found 1 target...
ERROR: /private/var/tmp/_bazel_irving/d4bad52af7e8ccc6c91f211db4181990/external/com_google_absl/absl/base/BUILD.bazel:118:1: undeclared inclusion(s) in rule '@com_google_absl//absl/base:base':
this rule is missing dependency declarations for the following files included by 'external/com_google_absl/absl/base/internal/unscaledcycleclock.cc':
  '/Library/Developer/CommandLineTools/SDKs/MacOSX10.13.sdk/usr/include/stdint.h'
  '/Library/Developer/CommandLineTools/SDKs/MacOSX10.13.sdk/usr/include/sys/_types/_int8_t.h'
  '/Library/Developer/CommandLineTools/SDKs/MacOSX10.13.sdk/usr/include/sys/_types/_int16_t.h'
  '/Library/Developer/CommandLineTools/SDKs/MacOSX10.13.sdk/usr/include/sys/_types/_int32_t.h'
  '/Library/Developer/CommandLineTools/SDKs/MacOSX10.13.sdk/usr/include/sys/_types/_int64_t.h'
  '/Library/Developer/CommandLineTools/SDKs/MacOSX10.13.sdk/usr/include/_types/_uint8_t.h'
  '/Library/Developer/CommandLineTools/SDKs/MacOSX10.13.sdk/usr/include/_types/_uint16_t.h'
  '/Library/Developer/CommandLineTools/SDKs/MacOSX10.13.sdk/usr/include/_types/_uint32_t.h'
  '/Library/Developer/CommandLineTools/SDKs/MacOSX10.13.sdk/usr/include/_types/_uint64_t.h'
  '/Library/Developer/CommandLineTools/SDKs/MacOSX10.13.sdk/usr/include/sys/_types.h'
  ...
Target //tensorflow/tools/pip_package:build_pip_package failed to build
Use --verbose_failures to see the command lines of failed build steps.
INFO: Elapsed time: 28.196s, Critical Path: 0.36s
INFO: 0 processes.
FAILED: Build did NOT complete successfully
```

**Comment by martinwicke** [comment]
@dslomov does this look like a bazel issue or an absl issue to you?

**Comment by girving** [comment]
@martinwicke It's not an `absl` thing: this time I got the error out of protobuf:

```
INFO: Analysed target //tensorflow/tools/pip_package:build_pip_package (332 packages loaded).
INFO: Found 1 target...
ERROR: /private/var/tmp/_bazel_irving/d4bad52af7e8ccc6c91f211db4181990/external/protobuf_archive/BUILD:105:1: undeclared inclusion(s) in rule '@protobuf_archive//:protobuf':
this rule is missing dependency declarations for the following files included by 'external/protobuf_archive/src/google/protobuf/util/internal/object_writer.cc':
  '/Library/Developer/CommandLineTools/SDKs/MacOSX10.13.sdk/usr/include/string.h'
  '/Library/Developer/CommandLineTools/SDKs/MacOSX10.13.sdk/usr/include/_types.h'
  '/Library/Developer/CommandLineTools/SDKs/MacOSX10.13.sdk/usr/include/sys/_types.h'
  '/Library/Developer/CommandLineTools/SDKs/MacOSX10.13.sdk/usr/include/sys/cdefs.h'
  '/Library/Developer/CommandLineTools/SDKs/MacOSX10.13.sdk/usr/include/sys/_symbol_aliasing.h'
  ...
```

**Comment by girving** [comment]
Unfortunately can't try with bazel HEAD since my currently broken version of bazel can't build bazel either. :)

**Comment by girving** [comment]
Reverting the Xcode command line tools back from `Xcode 10.0` to `Xcode 9.4.1` and rerunning `bazel clean --expunge` seems to dodge the problem, though bazel should probably still be fixed to work with Xcode 10.

**Comment by jgavris** [comment]
I don't think this is an issue of 'it doesn't work with Xcode 10', rather bazel does not gracefully handle switching Xcode versions / toolchains after it's done the initial Xcode configuration. Which is why it works fine after a clean.

**Comment by girving** [comment]
@jgavris It only works after a clean for Xcode 9.  It *does not* work after clean for Xcode 10.

**Comment by jgavris** [comment]
Oh! Interesting. I can confirm.

**Comment by cjolivier01** [comment]
I got this on absl and then later on protobuf like another was saying. Seems maybe related to which compiler bazel is choosing (I have several) and it's not clear which, or ho to control which it chooses

**Comment by jin** [comment]
@susinmotion since you've been working with Xcode version selections recently, do you have any idea about this issue?

**Comment by susinmotion** [comment]
Hm, does this problem persist after opening Xcode to finish the setup of command line tools, and then doing bazel shutdown so Bazel becomes aware of the new version?

**Comment by susinmotion** [comment]
Is this potentially a duplicate of https://github.com/bazelbuild/bazel/issues/2852?
Does setting `BAZEL_SH=/bin/bash` help?

**Comment by cjolivier01** [comment]
I think this got assigned to me on accident? or should I look into it?

**Comment by jmmv** [comment]
Based on the original report, this looks like you had the Command Line Tools installed but not Xcode, or `xcode-select` was pointing at the former.

Does this problem still exist?

(The ownership change was likely because this is now waiting on your input. I'm not sure if this workflow works well in GitHub though.)

**Comment by jmmv** [comment]
Looking through this bug's history again, I see:

> I don't think this is an issue of 'it doesn't work with Xcode 10', rather bazel does not gracefully handle switching Xcode versions / toolchains after it's done the initial Xcode configuration. Which is why it works fine after a clean.

Which is correct. Bazel has to cache Xcode information for performance reasons.

*However*, back when this bug was filed, the information was cached on disk and required an expensive `bazel clean --expunge` to fix. Today, we cache this information in memory so a `bazel shutdown` is sufficient to clear this incorrect state. This is the documented behavior now.

I'm not sure we easily can do better. (I can think of "stuff" like monitoring for changes in `/Applications` and triggering a rescan then... but that'd be fragile I suspect and I'm not sure the benefit is worth it.)

So... I think we should call this "WAI" unfortunately.

**Comment by paraboul** [comment]
Hey,

I just ran into this issue after I switched to Bazel 3.0.1 because of : https://github.com/bazelbuild/bazel/pull/8975

Basically, I recently removed Xcode, installed Command Line Tool :
```
$ xcode-select -p
$ /Library/Developer/CommandLineTools
```

```
$ xcrun --sdk macosx --show-sdk-path
$ /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk
```

Tried to clean, shutdown, etc.. 

```
 exec env - \
    PATH=/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/usr/local/go/bin \
    PWD=/proc/self/cwd \
  external/local_config_cc/cc_wrapper.sh -U_FORTIFY_SOURCE -fstack-protector -Wall -Wthread-safety -Wself-assign -fcolor-diagnostics -fno-omit-frame-pointer -g0 -O2 '-D_FORTIFY_SOURCE=1' -DNDEBUG -ffunction-sections -fdata-sections '-std=c++0x' -MD -MF bazel-out/darwin-opt/bin/external/skia/_objs/third_party/harfbuzz-harfbuzz/hb-subset-cff2.d '-frandom-seed=bazel-out/darwin-opt/bin/external/skia/_objs/third_party/harfbuzz-harfbuzz/hb-subset-cff2.o' -DHAVE_ICU_BUILTIN '-DU_USING_ICU_NAMESPACE=0' -DHAVE_OT -DHAVE_INTEL_ATOMIC_PRIMITIVES -DSK_USING_THIRD_PARTY_ICU -DHAVE_ICU -iquote external/skia -iquote bazel-out/darwin-opt/bin/external/skia -v '-std=c++17' -w -Wno-attributes -fstrict-aliasing -fPIC '-fvisibility=hidden' -isysroot /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk -Wno-unused-parameter -I external/skia//third_party/harfbuzz -I external/skia//third_party/externals/harfbuzz/src -I external/skia//third_party/externals/icu/source/common -I external/skia//third_party/externals/icu/source/i18n -I external/skia//third_party/icu -no-canonical-prefixes -Wno-builtin-macro-redefined '-D__DATE__="redacted"' '-D__TIMESTAMP__="redacted"' '-D__TIME__="redacted"' -c external/skia/third_party/externals/harfbuzz/src/hb-subset-cff2.cc -o bazel-out/darwin-opt/bin/external/skia/_objs/third_party/harfbuzz-harfbuzz/hb-subset-cff2.o)
ERROR: /private/var/tmp/_bazel_anthonycatel/0d9658930c3435936fb13a53bf4421d0/external/skia/BUILD.bazel:16:1: undeclared inclusion(s) in rule '@skia//:third_party/icu-icu':
this rule is missing dependency declarations for the following files included by 'external/skia/third_party/externals/icu/source/common/appendable.cpp':
  '/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/pthread.h'
  '/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/TargetConditionals.h'
  '/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/stdint.h'
  '/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/sys/_types/_int8_t.h'
  '/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/sys/_types/_int16_t.h'
  '/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/sys/_types/_int32_t.h'
  '/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/sys/_types/_int64_t.h'
  '/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/_types/_uint8_t.h'
  '/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/_types/_uint16_t.h'
  '/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/_types/_uint32_t.h'
  '/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/_types/_uint64_t.h'
  '/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/sys/_types.h'
  '/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/sys/cdefs.h'
  '/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/sys/_symbol_aliasing.h'
  '/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/sys/_posix_availability.h'
  '/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/machine/_types.h'
  '/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/i386/_types.h'
  '/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/sys/_pthread/_pthread_types.h'
  '/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/sys/_types/_intptr_t.h'
  '/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/machine/types.h'
  '/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/i386/types.h'
  '/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/sys/_types/_u_int8_t.h'
  '/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/sys/_types/_u_int16_t.h'
  '/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/sys/_types/_u_int32_t.h'
  '/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/sys/_types/_u_int64_t.h'
  '/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/sys/_types/_uintptr_t.h'
  '/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/_types/_intmax_t.h'
  '/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/_types/_uintmax_t.h'
  '/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/float.h'
Apple clang version 11.0.0 (clang-1100.0.33.17)
Target: x86_64-apple-darwin18.6.0
Thread model: posix
InstalledDir: /Library/Developer/CommandLineTools/usr/bin
 "/Library/Developer/CommandLineTools/usr/bin/clang" -cc1 -triple x86_64-apple-macosx10.14.0 -Wdeprecated-objc-isa-usage -Werror=deprecated-objc-isa-usage -emit-obj -disable-free -disable-llvm-verifier -discard-value-names -main-file-name appendable.cpp -mrelocation-model pic -pic-level 2 -mthread-model posix -mdisable-fp-elim -fno-strict-return -masm-verbose -munwind-tables -target-sdk-version=10.15 -target-cpu penryn -dwarf-column-info -debugger-tuning=lldb -ggnu-pubnames -target-linker-version 530 -v -ffunction-sections -fdata-sections -coverage-notes-file /private/var/tmp/_bazel_anthonycatel/0d9658930c3435936fb13a53bf4421d0/execroot/rave/bazel-out/darwin-opt/bin/external/skia/_objs/third_party/icu-icu/appendable.gcno -resource-dir /Library/Developer/CommandLineTools/usr/lib/clang/11.0.0 -dependency-file bazel-out/darwin-opt/bin/external/skia/_objs/third_party/icu-icu/appendable.d -skip-unused-modulemap-deps -MT bazel-out/darwin-opt/bin/external/skia/_objs/third_party/icu-icu/appendable.o -sys-header-deps -iquote external/skia -iquote bazel-out/darwin-opt/bin/external/skia -isysroot /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk -U _FORTIFY_SOURCE -D _FORTIFY_SOURCE=1 -D NDEBUG -D U_ENABLE_DYLOAD=0 -D U_STATIC_IMPLEMENTATION -D U_I18N_IMPLEMENTATION -D U_COMMON_IMPLEMENTATION -D U_USING_ICU_NAMESPACE=0 -D SK_USING_THIRD_PARTY_ICU -I external/skia//third_party/externals/icu/source/common -I external/skia//third_party/externals/icu/source/i18n -I external/skia//third_party/icu -D "__DATE__=\"redacted\"" -D "__TIMESTAMP__=\"redacted\"" -D "__TIME__=\"redacted\"" -stdlib=libc++ -internal-isystem /Library/Developer/CommandLineTools/usr/bin/../include/c++/v1 -O2 -Wall -Wthread-safety -Wself-assign -Wno-attributes -Wno-unused-parameter -Wno-builtin-macro-redefined -Wno-framework-include-private-from-public -Wno-atimport-in-framework-header -Wno-extra-semi-stmt -Wno-quoted-include-in-framework-header -w -std=c++17 -fdeprecated-macro -fdebug-compilation-dir /private/var/tmp/_bazel_anthonycatel/0d9658930c3435936fb13a53bf4421d0/execroot/rave -ferror-limit 19 -fmessage-length 0 -fvisibility hidden -fvisibility-inlines-hidden -stack-protector 1 -mdarwin-stkchk-strong-link -fblocks -fencode-extended-block-signature -fregister-global-dtors-with-atexit -fobjc-runtime=macosx-10.14.0 -fcxx-exceptions -fexceptions -fmax-type-align=16 -fdiagnostics-show-option -fcolor-diagnostics -vectorize-loops -vectorize-slp -o bazel-out/darwin-opt/bin/external/skia/_objs/third_party/icu-icu/appendable.o -x c++ external/skia/third_party/externals/icu/source/common/appendable.cpp
clang -cc1 version 11.0.0 (clang-1100.0.33.17) default target x86_64-apple-darwin18.6.0
ignoring nonexistent directory "/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/c++/v1"
ignoring nonexistent directory "/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/local/include"
ignoring nonexistent directory "/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/Library/Frameworks"
#include "..." search starts here:
 external/skia
 bazel-out/darwin-opt/bin/external/skia
#include <...> search starts here:
 external/skia//third_party/externals/icu/source/common
 external/skia//third_party/externals/icu/source/i18n
 external/skia//third_party/icu
 /Library/Developer/CommandLineTools/usr/bin/../include/c++/v1
 /Library/Developer/CommandLineTools/usr/lib/clang/11.0.0/include
 /Library/Developer/CommandLineTools/usr/include
 /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include
 /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/System/Library/Frameworks (framework directory)
End of search list.
Target //frontend/macos:xxx failed to build
```

Not sure how to debug that.
Thanks!

**Comment by jmmv** [comment]
The Command Line Tools aren't sufficient (I'm wondering why exactly though; never bothered to dig into the details). You need Xcode, and `xcode-select` should point at it.

**Comment by jin** [comment]
@paraboul can you try `bazel sync --configure` please? Also on `clean`, did you do it with `--expunge`?

**Comment by swang203** [comment]
> @paraboul can you try `bazel sync --configure` please? Also on `clean`, did you do it with `--expunge`?

I encounter the same problem,
the 
`bezel sync --configure` solve it. Thanks.

**Comment by jin** [comment]
This is most definitely because the `osx_cc_configure.bzl` is a repository rule that does not track underlying system changes, specifically to XCode installation directories and system headers.

`bazel sync --configure` forces the C++ toolchain repository rules to re-run.

**Comment by paraboul** [comment]
Hey, thanks for replying so quickly

> The Command Line Tools aren't sufficient (I'm wondering why exactly though; never bothered to dig into the details). You need Xcode, and `xcode-select` should point at it.

I've just updated to Catalina out of frustration and... that fixed it (without having to install xcode).
I can't confirm `bezel sync --configure` then!

**Comment by jmmv** [comment]
That's not going to work I think. There is an explicit cache for Xcode locations, which I moved from disk to memory:

https://github.com/bazelbuild/bazel/commit/3a292efa55075e020d60daeeb7566b7e2ea721c2

**Comment by paraboul** [comment]
> That's not going to work I think. There is an explicit cache for Xcode locations, which I moved from disk to memory:
> 
> [3a292ef](https://github.com/bazelbuild/bazel/commit/3a292efa55075e020d60daeeb7566b7e2ea721c2)

I was able to build my project after a full expunge. Isn't that because `/usr/bin/xcrun` properly returns the SDK from CLT location?

**Comment by jlbhdfsl** [comment]
Deleting /usr/local/Cellar/protobuf directory resolved a similar issue for me.
